File: src/ios/DogWalker/Presentation/Scenes/Dog/AddDogViewModel.swift
// Combine framework - Latest version
import Combine
// Foundation framework - Latest version
import Foundation

// Internal dependencies using relative paths
import "../../../Domain/UseCases/Dog/AddDogUseCase"
import "../../../Data/Repositories/DogRepository"
import "../../Common/Extensions/UIViewController+Alert"

// MARK: - Human Tasks
/*
Prerequisites:
1. Ensure AddDogUseCase is properly configured with DogRepository
2. Verify validation rules for dog profile data match business requirements
3. Configure proper error handling and user feedback mechanisms
4. Review accessibility requirements for form validation feedback
*/

/// ViewModel for handling the addition of new dogs to the system
/// Requirement: Dog Profile Management (1.3 Scope/Core Features/User Management)
/// Supports the creation of dog profiles with details like breed, age, and special needs
public class AddDogViewModel {
    
    // MARK: - Properties
    
    private let addDogUseCase: AddDogUseCase
    
    /// Published property for error message display
    @Published public var errorMessage: String?
    
    /// Published property for loading state
    @Published public var isLoading: Bool = false
    
    // MARK: - Initialization
    
    /// Initializes the AddDogViewModel with required dependencies
    /// - Parameter addDogUseCase: The use case for adding dogs to the system
    public init(addDogUseCase: AddDogUseCase) {
        self.addDogUseCase = addDogUseCase
    }
    
    // MARK: - Public Methods
    
    /// Handles the addition of a new dog to the system
    /// - Parameters:
    ///   - name: The name of the dog
    ///   - breed: The breed of the dog
    ///   - age: The age of the dog in years
    ///   - ownerId: The ID of the dog's owner
    public func addDog(name: String, breed: String, age: Int, ownerId: String) {
        // Set loading state
        isLoading = true
        errorMessage = nil
        
        // Validate input fields
        do {
            try validateInput(name: name, breed: breed, age: age, ownerId: ownerId)
            
            // Create Dog instance
            let dog = Dog(
                id: UUID().uuidString,
                name: name,
                breed: breed,
                age: age,
                ownerId: ownerId
            )
            
            // Execute add dog use case
            try addDogUseCase.execute(dog: dog)
            
            // Reset loading state on success
            isLoading = false
            
            // Log successful dog addition
            Logger.info("Successfully added dog: \(dog.name)")
            
        } catch let error as ValidationError {
            handleError(error)
        } catch let error as PersistenceError {
            handleError(error)
        } catch {
            handleError(error)
        }
    }
    
    // MARK: - Private Methods
    
    /// Validates the input fields for adding a new dog
    /// - Parameters:
    ///   - name: The name of the dog
    ///   - breed: The breed of the dog
    ///   - age: The age of the dog
    ///   - ownerId: The ID of the dog's owner
    /// - Throws: ValidationError if any field is invalid
    private func validateInput(name: String, breed: String, age: Int, ownerId: String) throws {
        // Validate name
        guard !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw ValidationError.invalidField(name: "name", message: "Dog name cannot be empty")
        }
        
        // Validate breed
        guard !breed.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw ValidationError.invalidField(name: "breed", message: "Dog breed cannot be empty")
        }
        
        // Validate age
        guard age >= 0 && age <= 30 else {
            throw ValidationError.invalidField(name: "age", message: "Dog age must be between 0 and 30 years")
        }
        
        // Validate owner ID
        guard !ownerId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw ValidationError.invalidField(name: "ownerId", message: "Owner ID cannot be empty")
        }
    }
    
    /// Handles errors that occur during dog addition
    /// - Parameter error: The error to handle
    private func handleError(_ error: Error) {
        isLoading = false
        
        // Set appropriate error message based on error type
        if let validationError = error as? ValidationError {
            errorMessage = validationError.errorDescription
        } else if let persistenceError = error as? PersistenceError {
            errorMessage = "Failed to save dog: \(persistenceError.localizedDescription)"
        } else {
            errorMessage = "An unexpected error occurred: \(error.localizedDescription)"
        }
        
        // Log the error
        Logger.error("Error adding dog: \(error.localizedDescription)")
    }
}

File: src/android/app/src/main/java/com/dogwalker/app/presentation/screens/dog/AddDogViewModel.kt
package com.dogwalker.app.presentation.screens.dog

import androidx.lifecycle.ViewModel // androidx.lifecycle:lifecycle-viewmodel-ktx:2.6.1
import androidx.lifecycle.viewModelScope // androidx.lifecycle:lifecycle-viewmodel-ktx:2.6.1
import com.dogwalker.app.domain.model.Dog
import com.dogwalker.app.domain.usecase.dog.AddDogUseCase
import dagger.hilt.android.lifecycle.HiltViewModel // dagger.hilt:hilt-android:2.44
import kotlinx.coroutines.flow.MutableStateFlow // org.jetbrains.kotlinx:kotlinx-coroutines-core:1.6.4
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Human Tasks:
 * 1. Verify that Hilt dependency injection is properly configured in the application
 * 2. Ensure proper error handling is implemented in the UI layer when calling ViewModel methods
 * 3. Consider implementing unit tests to verify validation logic and state management
 * 4. Review error messages for user-friendliness and localization needs
 */

/**
 * ViewModel responsible for managing the state and business logic for adding new dog profiles.
 *
 * Requirements addressed:
 * - Dog Profile Management (1.3 Scope/Core Features/User Management)
 *   Implements the presentation layer logic for adding new dog profiles, managing UI state,
 *   and coordinating with the domain layer through the AddDogUseCase.
 */
@HiltViewModel
class AddDogViewModel @Inject constructor(
    private val addDogUseCase: AddDogUseCase
) : ViewModel() {

    /**
     * Sealed class representing the different states of the add dog operation
     */
    sealed class AddDogState {
        object Idle : AddDogState()
        object Loading : AddDogState()
        object Success : AddDogState()
        data class Error(val message: String) : AddDogState()
    }

    private val _state = MutableStateFlow<AddDogState>(AddDogState.Idle)
    val state: StateFlow<AddDogState> = _state.asStateFlow()

    /**
     * Handles the addition of a new dog profile.
     *
     * Requirements addressed:
     * - Dog Profile Management (1.3 Scope/Core Features/User Management)
     *   Implements the logic for validating and adding new dog profiles through
     *   the AddDogUseCase while managing the UI state.
     *
     * @param dog The Dog object containing the profile information to be added
     */
    fun addDog(dog: Dog) {
        viewModelScope.launch {
            try {
                _state.value = AddDogState.Loading

                // Additional view model level validation
                validateDog(dog)

                // Execute the use case to add the dog
                addDogUseCase.execute(dog)

                _state.value = AddDogState.Success
            } catch (e: IllegalArgumentException) {
                _state.value = AddDogState.Error(e.message ?: "Invalid dog profile data")
            } catch (e: Exception) {
                _state.value = AddDogState.Error("Failed to add dog profile: ${e.message}")
            }
        }
    }

    /**
     * Validates the dog profile data before attempting to add it.
     *
     * @param dog The Dog object to validate
     * @throws IllegalArgumentException if validation fails
     */
    private fun validateDog(dog: Dog) {
        require(dog.name.length <= Dog.MAX_NAME_LENGTH) {
            "Dog name cannot exceed ${Dog.MAX_NAME_LENGTH} characters"
        }

        require(dog.age in Dog.MIN_AGE..Dog.MAX_AGE) {
            "Dog age must be between ${Dog.MIN_AGE} and ${Dog.MAX_AGE} years"
        }

        require(dog.breed.isNotBlank()) {
            "Dog breed cannot be blank"
        }

        require(dog.ownerId.isNotBlank()) {
            "Owner ID cannot be blank"
        }
    }

    /**
     * Resets the state to idle.
     * Useful when navigating away from the add dog screen or when preparing for a new operation.
     */
    fun resetState() {
        _state.value = AddDogState.Idle
    }
}